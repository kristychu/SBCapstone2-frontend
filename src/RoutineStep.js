import React, { useState, useEffect } from "react";
import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from "reactstrap";
import ProductSearch from "./ProductSearch";
import { Link } from "react-router-dom";
import { getProductDetails, deleteProductFromStep } from "./actions";
import { useSelector, useDispatch } from "react-redux";

function RoutineStep({ routineStep, productId, timeOfDay, id }) {
  const token = useSelector((st) => st.users.profile.token);
  const dispatch = useDispatch();
  const products = useSelector((st) => st.products);
  let productBrand;
  let productName;
  if (products) {
    const savedProducts = Object.values(products);
    for (let p of savedProducts) {
      if (p.id === productId) {
        productBrand = p.productBrand;
        productName = p.productName;
      }
    }
  }

  //Modal functionality
  const [modal, setModal] = useState(false);
  const toggle = () => setModal(!modal);

  useEffect(() => {
    if (productId) {
      console.log("API CALL FOR PRODUCT INFO!");
      dispatch(getProductDetails(productId));
    }
  }, [dispatch, productId]);
  if (!token) {
    return (
      <div>
        <p>You need an account to access this page.</p>
        <p>
          Have an account already? <Link to="/login">Login!</Link>
        </p>
        <p>
          Otherwise, <Link to="/register">Register here</Link>.
        </p>
      </div>
    );
  } else {
    const decoded = JSON.parse(atob(token.split(".")[1]));
    const username = decoded.username;

    const handleDelete = () => {
      dispatch(
        deleteProductFromStep(username, token, id, routineStep, timeOfDay)
      );
    };

    return (
      <>
        <td key={id}>
          {productId && (
            <>
              <p>
                {productBrand} {productName}
                <Button color="danger" onClick={handleDelete}>
                  Delete
                </Button>
              </p>
            </>
          )}
          {!productId && (
            <Button color="info" onClick={toggle}>
              Find Product
            </Button>
          )}
        </td>
        <Modal isOpen={modal} toggle={toggle}>
          <ModalHeader toggle={toggle}>Search for {routineStep}</ModalHeader>
          <ModalBody>
            <ProductSearch
              routineStep={routineStep}
              timeOfDay={timeOfDay}
              toggle={toggle}
            />
          </ModalBody>
          <ModalFooter></ModalFooter>
        </Modal>
      </>
    );
  }
}

export default RoutineStep;
