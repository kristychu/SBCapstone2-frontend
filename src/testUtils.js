import thunk from "redux-thunk";
import configureStore from "redux-mock-store";

export const makeMockStore = configureStore([thunk]);
