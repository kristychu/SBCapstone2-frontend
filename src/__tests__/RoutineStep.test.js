import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import reducer from "../reducers/rootReducer";
import { render } from "@testing-library/react";
import RoutineStep from "../RoutineStep";
import { MemoryRouter } from "react-router";
import { Table } from "reactstrap";
import { makeMockStore } from "../testUtils";

//USING ACTUAL REDUCER WHICH SHOULD BE EMPTY
const store = createStore(reducer);

it("should render without crashing", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <RoutineStep />
      </MemoryRouter>
    </Provider>
  );
});

it("should match snapshot for anonymous user", function () {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <RoutineStep />
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

// CREATING A MOCK STORE AND FAKED DATA
const mockStore = makeMockStore({
  users: {
    profile: {
      token:
        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InRlc3R1c2VyIiwiaWF0IjoxNjE4NDQzNDQ0fQ.8PZOXva_Bztp48k8obLY5sj8k7siJVMTIhftj6509EI",
    },
  },
  steps: [
    [
      {
        stepNumber: 0,
        routineStep: "Makeup Remover and Oil Cleanser",
        timeOfDay: "morning",
        productId: null,
        stepId: null,
      },
      {
        stepNumber: 1,
        routineStep: "Water Based Cleanser",
        timeOfDay: "morning",
        productId: null,
        stepId: null,
      },
      {
        stepNumber: 2,
        routineStep: "Exfoliator",
        timeOfDay: "morning",
        productId: null,
        stepId: null,
      },
    ],
    [
      {
        stepNumber: 11,
        routineStep: "Makeup Remover and Oil Cleanser",
        timeOfDay: "night",
        productId: null,
        stepId: null,
      },
      {
        stepNumber: 12,
        routineStep: "Water Based Cleanser",
        timeOfDay: "night",
        productId: null,
        stepId: null,
      },
      {
        stepNumber: 13,
        routineStep: "Exfoliator",
        timeOfDay: "night",
        productId: null,
        stepId: null,
      },
    ],
  ],
});

it("should match snapshot if logged in", function () {
  const { asFragment } = render(
    <Provider store={mockStore}>
      <MemoryRouter>
        <Table>
          <tbody>
            <tr>
              <RoutineStep />
            </tr>
          </tbody>
        </Table>
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should have a find button if logged in and user does not have anything saved in database", function () {
  const { getByText, queryByText } = render(
    <Provider store={mockStore}>
      <MemoryRouter>
        <Table>
          <tbody>
            <tr>
              <RoutineStep />
            </tr>
          </tbody>
        </Table>
      </MemoryRouter>
    </Provider>
  );

  expect(getByText("Find Product")).toBeInTheDocument();
  expect(queryByText("Delete")).toBeFalsy();
});
