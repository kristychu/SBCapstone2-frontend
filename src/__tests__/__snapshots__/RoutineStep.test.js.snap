// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot for anonymous user 1`] = `
<DocumentFragment>
  <div>
    <p>
      You need an account to access this page.
    </p>
    <p>
      Have an account already? 
      <a
        href="/login"
      >
        Login!
      </a>
    </p>
    <p>
      Otherwise, 
      <a
        href="/register"
      >
        Register here
      </a>
      .
    </p>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot if logged in 1`] = `
<DocumentFragment>
  <table
    class="table"
  >
    <tbody>
      <tr>
        <td>
          <button
            class="btn btn-info"
            type="button"
          >
            Find Product
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
